tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetAsFunRule.scala

```diff
@@ -28,12 +28,63 @@ class SetAsFunRule(rewriter: SymbStateRewriter) extends RewritingRule {
         val setCell = nextState.asCell
         setEx.typeTag match {
           case Typed(SetT1(TupT1(keyType, valueType))) =>
-            nextState = translateRelation(setCell, nextState)
-            val rel = nextState.asCell
             // construct a cell for the function and attach the relation to it
             nextState = nextState.updateArena(_.appendCell(CellT.fromType1(FunT1(keyType, valueType))))
             val fun = nextState.arena.topCell
-            nextState = nextState.updateArena(_.setCdm(fun, rel))
+
+            rewriter.solverContext.config.smtEncoding match {
+              case `arraysEncoding` =>
+                nextState = nextState.updateArena(_.appendCell(FinSetT(CellT.fromType1(keyType))))
+                val domainCell = nextState.arena.topCell
+                nextState = nextState.updateArena(_.appendCellNoSmt(setCell.cellType))
+                val relationCell = nextState.arena.topCell
+
+                val pairs = nextState.arena.getHas(setCell)
+                var domainCells: List[ArenaCell] = List()
+                var rangeCells: List[ArenaCell] = List()
+
+                for (pair <- pairs) {
+                  assert(nextState.arena.getHas(pair).length == 2)
+                  val domElem = nextState.arena.getHas(pair)(0)
+                  val resElem = nextState.arena.getHas(pair)(1)
+
+                  nextState = nextState.updateArena(_.appendHas(domainCell, domElem))
+                  val inExpr = OperEx(ApalacheOper.storeInSet, domElem.toNameEx, domainCell.toNameEx)
+                  rewriter.solverContext.assertGroundExpr(inExpr)
+
+                  nextState = nextState.updateArena(_.appendHasNoSmt(relationCell, pair))
+
+                  domainCells = domainCells :+ domElem
+                  rangeCells = rangeCells :+ resElem
+                }
+
+                nextState = nextState.updateArena(_.setDom(fun, domainCell))
+                nextState = nextState.updateArena(_.setCdm(fun, relationCell))
+
+                def addCellCons(domElem: ArenaCell, rangeElem: ArenaCell): Unit = {
+                  val inDomain = tla.apalacheSelectInFun(domElem.toNameEx, domainCell.toNameEx)
+                  val inRange =
+                    tla.apalacheStoreInFun(rangeElem.toNameEx, fun.toNameEx, domElem.toNameEx)
+                  val notInRange = tla.apalacheStoreNotInFun(domElem.toNameEx, fun.toNameEx)
+                  // function updates are guarded by the inDomain predicate
+                  val ite = tla.ite(inDomain, inRange, notInRange)
+                  rewriter.solverContext.assertGroundExpr(ite)
+                }
+
```

Comment:
I am not sure about what happens when `SetAsFun({ <<0, 1>>, <<0, 2>> })` is called

Reply:
We should have `f(0) = 1`. Following the way it is implemented for the oopsla19 encoding, for duplicate keys we use the "first" pair. Odd to mention order when dealing with sets, but we simply use the key that is iterated over first.